### PR TITLE
Backports 0.18 pr16

### PR DESCRIPTION
**Pull request by Munkybooty**
wallet_dump.py keeps failiing but everything else works

### CODE REVIEW

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]

**Comment by PastaPastaPasta** [comment]
(this wasn't a full review) ^

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
`wallet_dump.py` failure is somehow related to 15415 it seem but I don't quite understand how, might want to just drop it for now

File: contrib/init/dashd.service [hunk 1]
```diff
@@ -5,21 +5,45 @@
...
 Type=forking
 PIDFile=/run/dashd/dashd.pid
 Restart=on-failure
 
+# Directory creation and permissions
+####################################
+
+# Run as dash:dash
+User=dash
+Group=dash
```
Comment (UdjinM6) [inline contrib/init/dashd.service:33]:
```suggestion
User=dashcore
Group=dashcore
```

File: doc/init.md [hunk 2]
```diff
@@ -65,6 +65,22 @@ reasons to make the configuration file and data directory only readable by the
...
+dashcore user and group to do so (e.g. when `-sysperms` is specified). This does not allow
+for the listing of files under the directory.
+
+NOTE: It is not currently possible to override `datadir` in
+`/etc/dash/dash.conf` with the current systemd, OpenRC, and Upstart init
+files out-of-the-box. This is because the command line options specified in the
+init files take precedence over the configurations in
+`/etc/dash/dash.conf`. However, some init systems have their own
+configuration mechanisms that would allow for overriding the command line
+options specified in the init files (e.g. setting `DASHD_DATADIR` for
```
Comment (UdjinM6) [inline doc/init.md:81]:
```suggestion
options specified in the init files (e.g. setting `BITCOIND_DATADIR` for
```

NOTE: we should dashify `BITCOIND_*` env vars maybe (in another PR)

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -1726,7 +1726,7 @@ static void MaybePushAddress(UniValue & entry, const CTxDestination &dest)
  * @param  ret        The UniValue into which the result is stored.
  * @param  filter     The "is mine" filter bool.
  */
-static void ListTransactions(CWallet * const pwallet, const CWalletTx& wtx, const std::string& strAccount, int nMinDepth, bool fLong, UniValue& ret, const isminefilter& filter) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+static void ListTransactions(CWallet * const pwallet, const CWalletTx& wtx, const std::string& strAccount, int nMinDepth, bool fLong, UniValue& ret, const isminefilter& filter) EXCLUSIVE_LOCKS_REQUIRED(cs_main) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet)
```
Comment (PastaPastaPasta) [inline src/wallet/rpcwallet.cpp:1729]: 15365 `CWallet * const pwallet` -> `CWallet* const pwallet`

Also, @UdjinM6 should these two EXCLUSIVE_LOCKS_REQUIREDs be combined?

File: src/test/txvalidationcache_tests.cpp [hunk 4]
```diff
@@ -114,7 +114,7 @@ BOOST_FIXTURE_TEST_CASE(tx_mempool_block_doublespend, TestChain100Setup)
 // should fail.
 // Capture this interaction with the upgraded_nop argument: set it when evaluating
 // any script flag that is implemented as an upgraded NOP code.
-static void ValidateCheckInputsForAllFlags(const CTransaction &tx, uint32_t failing_flags, bool add_to_cache)
+static void ValidateCheckInputsForAllFlags(const CTransaction &tx, uint32_t failing_flags, bool add_to_cache) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
```
Comment (PastaPastaPasta) [inline src/test/txvalidationcache_tests.cpp:117]: 15413 seems incomplete

File: test/functional/wallet_abandonconflict.py [hunk 5]
```diff
@@ -24,8 +24,7 @@
 class AbandonConflictTest(BitcoinTestFramework):
     def set_test_params(self):
         self.num_nodes = 2
-        # TODO: remove -txindex. Currently required for getrawtransaction call.
-        self.extra_args = [["-minrelaytxfee=0.00001", "-txindex"], []]
+        self.extra_args = [["-minrelaytxfee=0.00001"], []]
```
Comment (PastaPastaPasta) [inline test/functional/wallet_abandonconflict.py:27]: 15404 missing changes here https://github.com/dashpay/dash/blob/8226f22efb1dd72d89fab1897b83a9239759f092/test/functional/wallet_basic.py#L25-L26

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]

File: src/wallet/rpcwallet.cpp (hunk 3 above)
Comment (UdjinM6) [reply src/wallet/rpcwallet.cpp:1729]: Good catch!

File: src/test/txvalidationcache_tests.cpp (hunk 4 above)
Comment (Munkybooty) [reply src/test/txvalidationcache_tests.cpp:117]: I've looked through 15413 and i can't seem to find what would be missing.

Comment (PastaPastaPasta) [reply src/test/txvalidationcache_tests.cpp:117]: I think I was referring to the bench/* changes. Which means we should probably mark this as partial, or drop until block_assemble and duplicate_inputs exsists

Comment (Munkybooty) [reply src/test/txvalidationcache_tests.cpp:117]: Sounds good

**Review by UdjinM6 — APPROVED** [review]
LGTM 👍 


utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merging via merge commit

**Pull request merged**